### DISCUSSION TITLE
Linux: Raspberry Pi OS (aka Raspbian)

### DISCUSSION THREAD

**Discussion by ylluminate** [vlang/v#13759] [Installation Issues]
Please comment with or ask questions about your experience or trouble installing V on the [Raspberry Pi OS](https://en.wikipedia.org/wiki/Raspberry_Pi_OS) (formerly known as Raspbian) Linux distribution for the Raspberry Pi. 

Official website: https://www.raspberrypi.com/software/operating-systems/

**Comment by larpon** [comment]
Here's a few of my experiences with V @ version `0.2.4 8a2d3f6` ... `V 0.4.3 e0207b6`

The following worked for me on Raspian 10 (buster) on a rPi4 / Raspberry Pi 4 (model B)

The following `apt` packages seem to be the minimum required to get V bootstrapped:

`sudo apt-get install git build-essential libatomic1 libgc-dev libc6-dev`

gcc
---
(`gcc` is usually the fallback compiler if `tcc` fails - see about `tcc` below)
If V uses `gcc` but [fails with something like](https://github.com/vlang/v/issues/12907):
```
/usr/bin/ld: /tmp/ccplBfxf.o: in function `atomic_load_u64':
v.c:(.text+0xa78): undefined reference to `__atomic_load_8'
/usr/bin/ld: /tmp/ccplBfxf.o: in function `atomic_store_u64':
v.c:(.text+0xad0): undefined reference to `__atomic_store_8'
/usr/bin/ld: /tmp/ccplBfxf.o: in function `atomic_fetch_add_u64':
v.c:(.text+0xb04): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: /tmp/ccplBfxf.o: in function `atomic_fetch_sub_u64':
v.c:(.text+0xb44): undefined reference to `__atomic_fetch_sub_8'
collect2: error: ld returned 1 exit status
```

You can try something like:
```bash
LDFLAGS=-latomic v ...
```
or to bootstrap
```bash
export LDFLAGS=-latomic
make
```

tcc
---
To get `tcc` working I had to compile it manually first, fix a library lookup problem and add `LDFLAGS=-latomic` like with `gcc`.
I ended up with [this little script](https://gist.github.com/Larpon/998febd5f3c33f995f0ecd561c043dcb):
```bash
#!/bin/bash

## should be run in V's main repo folder!

# Build tcc.exe

rm -rf tinycc/
rm -rf thirdparty/tcc/

pushd .

git clone git://repo.or.cz/tinycc.git
cd tinycc

./configure \
            --prefix=thirdparty/tcc \
            --bindir=thirdparty/tcc \
            --crtprefix=thirdparty/tcc/lib:/usr/lib:/usr/lib/arm-linux-gnueabihf \
            --libpaths=thirdparty/tcc/lib:/usr/lib/arm-linux-gnueabihf:/usr/lib:/lib/arm-linux-gnueabihf:/lib:/usr/local/lib/arm-linux-gnueabihf:/usr/local/lib \
            --extra-cflags="-O3 -flto" \
            --debug
            
make
make install

popd

mv tinycc/thirdparty/tcc thirdparty/tcc

mv thirdparty/tcc/tcc thirdparty/tcc/tcc.exe

thirdparty/tcc/tcc.exe -v -v

# Make tcc able to find `libatomic`:

ln -s /usr/lib/arm-linux-gnueabihf/libatomic.so.1 "$(pwd)/thirdparty/tcc/lib/libatomic.so"
ln -s /usr/lib/arm-linux-gnueabihf/libgc.a "$(pwd)/thirdparty/tcc/lib/libgc.a"

# Tell compiler to link against `libatomic`:
export LDFLAGS=-latomic
```
The build process is taken directly from the [vlang/tccbin arm branch](https://github.com/vlang/tccbin/tree/thirdparty-linux-arm#thirdparty-linux-arm)

**Comment by kake26** [comment]
Here are a few key notes on building V on a Raspberry Pi. 

These notes pertain to V 0.3.2 on a Raspberry Pi 4 running Raspberry Pi OS 11 (bullseye). I run the 32 bit version, however these should apply to 64 bit as well. 

First, I recommend to clone the V github repo as I've had issues with building the release versions. That said there is one small change you need to make to the GNUMakefile for V to build via its normal process. This is a different application of essentially the same thing noted in the previous comment here by Larpon. 

**GNUMakefile**

I change

`LDFLAGS ?=`

To

`LDFLAGS += -latomic`

V should build normally after this change.

**Reply by ylluminate** [reply]
When building apps with V on your Pi with this build you also used the following for a little more optimization with success for deployment to your more limited router(s), right?: 
```sh
export LDFLAGS='-latomic'
export VFLAGS='-cc gcc -compress -skip-unused -prod'
```

**Reply by kake26** [reply]
Correct. Though -compress needs the system to have UPX installed. You can install that via apt easily. Also I set my LDFLAGS to -static for maximum portability. I'll make sure to detail my full setup for building things that will run on a router in a separate post.
